kernel: usb: fix dependencies for gadget drivers

thom24 opened the pull request:
Fix dependencies for gadget drivers. The gadget packages do not need to depends on kmod-usb-core.

Fix also some controllers to be able to build them in gadget only mode, or dual-role mode.

thom24 committed:
kernel: usb: kmod-usb-dwc3: add gadget and dual-role modes support

Build DWC3 driver in the right mode:
- host mode if kernel have usb host support only.
- gadget mode if kernel have usbgadget support only.
- dual-role mode if kernel have both usb host and usbgadget support.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/19246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: usb: gadget drivers do not depend on usb host support

The usb gadget drivers don't depend on usb host support, so fix the package
dependencies accordingly.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/19246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: usb: fix dependencies for kmod-usb-roles package

The usb role driver does not depend on the kernel config CONFIG_USB (host
side support). It only depends on the kernel config CONFIG_USB_SUPPORT.

Now it depends on the OpenWrt configurations USB_SUPPORT or
USB_GADGET_SUPPORT (usb and usbgadget features). We can assume that if usb
or usbgadget features are set, CONFIG_USB_SUPPORT has been set in the
target/subtarget kernel config.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/19246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 committed:
kernel: usb: fix dependencies for kmod-cdns3 and kmod-cdns

The cdns3 driver can be built without usb host mode support. Fix
dependencies to handle the case it is built with only gadget support.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/19246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 commented on package/kernel/linux/modules/usb.mk:
```diff
@@ -235,12 +234,12 @@ endef
+  SUBMENU:=$(USB_MENU)
   TITLE:= USB CDC Composite (Ethernet + ACM)
   KCONFIG:=CONFIG_USB_CDC_COMPOSITE
   DEPENDS:=+kmod-usb-gadget +kmod-usb-lib-composite \
```
It is not different, I just missed it

hauke commented on package/kernel/linux/modules/usb.mk:
```diff
@@ -498,17 +500,26 @@ $(eval $(call KernelPackage,usb-cdns3))
+	   +USB_SUPPORT:kmod-usb-core \
+	   +kmod-usb-roles
   KCONFIG:= \
 	CONFIG_USB_DWC3 \
-	CONFIG_USB_DWC3_HOST=y \
+ifeq($(CONFIG_USB_SUPPORT)$(CONFIG_USB_GADGET_SUPPORT),yy)
+	CONFIG_USB_DWC3_HOST=n \
 	CONFIG_USB_DWC3_GADGET=n \
+	CONFIG_USB_DWC3_DUAL_ROLE=y \
+else
+	CONFIG_USB_DWC3_HOST=$(if $(CONFIG_USB_SUPPORT),y,n) \
+	CONFIG_USB_DWC3_GADGET=$(if $(CONFIG_USB_GADGET_SUPPORT),y,n) \
 	CONFIG_USB_DWC3_DUAL_ROLE=n \
+endif
 	CONFIG_USB_DWC3_DEBUG=n \
 	CONFIG_USB_DWC3_VERBOSE=n
```
This is not working.

The build of multiple targets failed like this:
```
   DesignWare USB3 DRD Core Support (USB_DWC3) [M/n/y/?] m
    Register ULPI PHY Interface (USB_DWC3_ULPI) [N/y/?] n
    DWC3 Mode Selection
      1. Host only mode (USB_DWC3_HOST) (NEW)
      2. Gadget only mode (USB_DWC3_GADGET) (NEW)
    > 3. Dual Role mode (USB_DWC3_DUAL_ROLE) (NEW)
    choice[1-3?]: make[7]: *** [scripts/kconfig/Makefile:85: syncconfig] Error 1
make[6]: *** [Makefile:686: syncconfig] Error 2
```

I think you have to change it like this:
```
  KCONFIG:= \
	CONFIG_USB_DWC3 \
	CONFIG_USB_DWC3_HOST=y \
	CONFIG_USB_DWC3_DEBUG=n \
	CONFIG_USB_DWC3_VERBOSE=n
ifeq($(CONFIG_USB_SUPPORT)$(CONFIG_USB_GADGET_SUPPORT),yy)
  KCONFIG+= \
	CONFIG_USB_DWC3_HOST=n \
	CONFIG_USB_DWC3_GADGET=n \
	CONFIG_USB_DWC3_DUAL_ROLE=y
else
  KCONFIG+= \
	CONFIG_USB_DWC3_HOST=$(if $(CONFIG_USB_SUPPORT),y,n) \
	CONFIG_USB_DWC3_GADGET=$(if $(CONFIG_USB_GADGET_SUPPORT),y,n) \
	CONFIG_USB_DWC3_DUAL_ROLE=n
endif
```

hauke commented on package/kernel/linux/modules/usb.mk:
```diff
@@ -498,17 +500,28 @@ $(eval $(call KernelPackage,usb-cdns3))
-	CONFIG_USB_DWC3_DUAL_ROLE=n \
 	CONFIG_USB_DWC3_DEBUG=n \
 	CONFIG_USB_DWC3_VERBOSE=n
+ifeq($(CONFIG_USB_SUPPORT)$(CONFIG_USB_GADGET_SUPPORT),yy)
```
```suggestion
ifeq ($(CONFIG_USB_SUPPORT)$(CONFIG_USB_GADGET_SUPPORT),yy)
```
The space is important

hauke commented on package/kernel/linux/modules/usb.mk:
```diff
@@ -460,13 +460,15 @@ $(eval $(call KernelPackage,usb-dwc2-pci))
 
 
 define KernelPackage/usb-cdns
+  SUBMENU:=(USB_MENU)
```
```suggestion
  SUBMENU:=$(USB_MENU)
```

hauke commented:
The CI found an other problem:
```
Package kmod-usb-gadget is missing dependencies for the following libraries:
usb-common.ko
make[2]: *** [modules/usb.mk:107: /__w/openwrt/openwrt/openwrt/bin/targets/siflower/sf21/packages/kmod-usb-gadget-6.6.100-r1.apk] Error 1
```

thom24 committed:
kernel: usb: add kmod-usb-common package

Move usb-common driver from usb-core package to a new usb-common package.
The usb-common driver is needed by usb-gadget driver which can be used
without host support.

Signed-off-by: Thomas Richard <thomas.richard@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/19246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

thom24 commented:
Remaining CI errors seem not to be related to my changes.
But now I have a recursive dependency error
```
tmp/.config-package.in:24587:error: recursive dependency detected!
tmp/.config-package.in:24587:	symbol PACKAGE_kmod-brcmfmac depends on BRCMFMAC_USB
tmp/.config-package.in:24622:	symbol BRCMFMAC_USB depends on PACKAGE_kmod-brcmfmac
```

thom24 commented:
@hauke, can you have a look ?
CI errors are false positive.
